### PR TITLE
feat: add security provider request to AbstractMessageManager

### PR DESCRIPTION
**Pull request by vinistevam**
**Add security provider request to AbstractMessageManager**

**Description**

This PR aims to move the security provider request from `SignController` in the extension to `AbstractMessageManager`, and support it in a generic way here in the core and potentially in mobile. 

NOTE: There will be a following PR for the extension to delete the duplicated code within the extension, and potentially following that a PR moving the  `securityProviderCheck` from the extension to core utils.

See the patched changes in the extension (ci): [branch](https://github.com/MetaMask/metamask-extension/tree/poc/security-provider) 

- CHANGED:

  - Included a new param `securityProviderRequest` 
  - Created a method `securityProviderCheck` to be called whenever an unapproved message is added if securityProviderRequest is defined.
  - added a call to the `securityProviderCheck`  into the `addMessage` whenever security provided is provided, the method will check if the Message is malicious or not. 
  - Changed `addMessage` and `addUnapprovedMessage` to asynchronous 

- ADDED:

  - Unit test to cover the changes to support the security provider

**Checklist**

- [x] Tests are included if applicable
- [x] Any added code is fully documented

**Issue**

Resolves https://github.com/MetaMask/MetaMask-planning/issues/391

### CODE REVIEW

File: packages/message-manager/src/AbstractMessageManager.ts [hunk 1]
```diff
@@ -179,13 +195,26 @@ export abstract class AbstractMessageManager<
...
-   * the unapproved Messages from that list to this.messages.
+   * the unapproved Messages from that list to this.messages. If a security provider
+   * is provided, the method will check if the Message is malicious or not.
    *
    * @param message - The Message to add to this.messages.
    */
   addMessage(message: M) {
-    this.messages.push(message);
-    this.saveMessageList();
+    if (this.securityProviderRequest) {
```
Comment (matthewwalsh0) [inline packages/message-manager/src/AbstractMessageManager.ts:204]: For the sake of clarity, could this go in a separate private method like `checkSecurity` which checks if a request was provided and simply returns the updated message, or the original one.

File: packages/message-manager/src/AbstractMessageManager.ts [hunk 2]
```diff
@@ -179,13 +195,26 @@ export abstract class AbstractMessageManager<
...
+      this.securityProviderRequest(message, message.type).then(
+        (securityProviderResponse) => {
+          this.messages.push({
```
Comment (matthewwalsh0) [inline packages/message-manager/src/AbstractMessageManager.ts:207]: No need to duplicate this, we could simply always use the returned message from the new `checkSecurity` method and push and save once.

File: packages/message-manager/src/AbstractMessageManager.ts [hunk 3]
```diff
@@ -23,13 +24,15 @@ export interface OriginalRequest {
...
  * @property rawSig - Raw data of the signature request
+ * @property securityProviderResponse - Response from a security provider, whether it is malicious or not
  */
 export interface AbstractMessage {
   id: string;
   time: number;
   status: string;
   type: string;
   rawSig?: string;
+  securityProviderResponse?: Json;
```
Comment (matthewwalsh0) [inline packages/message-manager/src/AbstractMessageManager.ts:35]: As we know this is always an object, would `Map<string, Json>` be more accurate?

File: packages/message-manager/src/AbstractMessageManager.ts (hunk 2 above)
Comment (matthewwalsh0) [reply packages/message-manager/src/AbstractMessageManager.ts:207]: I see now this was likely to delay the side effects of pushing and saving until the async work was done.

As mobile and extension appear to only use `addUnapprovedMessageAsync` which is already returning a promise, one option is to make `addUnapprovedMessage` and `addMessage` also return promises so we can `await` them and simplify the flow.

The suffix `Async` then gets a bit confusing as it doesn't mean it's a synchronous method, but rather it doesn't wait for the message to be processed.

File: packages/message-manager/src/AbstractMessageManager.ts (hunk 1 above)
Comment (vinistevam) [reply packages/message-manager/src/AbstractMessageManager.ts]: Indeed make sense. I've added the `checkSecurity` and also the necessary tests.

File: packages/message-manager/src/AbstractMessageManager.ts (hunk 3 above)
Comment (vinistevam) [reply packages/message-manager/src/AbstractMessageManager.ts]: yes, I just pushed a fix.

**Review by bergarces — APPROVED** [review]

**Review by OGPoyraz — APPROVED** [review]

File: packages/message-manager/src/MessageManager.ts [hunk 4]
```diff
@@ -110,7 +110,10 @@ export class MessageManager extends AbstractMessageManager<
    * @param req - The original request object possibly containing the origin.
    * @returns The id of the newly created message.
    */
-  addUnapprovedMessage(messageParams: MessageParams, req?: OriginalRequest) {
+  async addUnapprovedMessage(
```
Comment (Gudahtt) [inline packages/message-manager/src/MessageManager.ts:113]: Looks like this is a breaking change (just noting this for the changelog)

File: packages/message-manager/src/PersonalMessageManager.ts [hunk 5]
```diff
@@ -113,10 +113,10 @@ export class PersonalMessageManager extends AbstractMessageManager<
    * @param req - The original request object possibly containing the origin.
    * @returns The id of the newly created message.
    */
-  addUnapprovedMessage(
+  async addUnapprovedMessage(
```
Comment (Gudahtt) [inline packages/message-manager/src/PersonalMessageManager.ts:116]: Looks like this is a breaking change (just noting this for the changelog)

File: packages/message-manager/src/TypedMessageManager.ts [hunk 6]
```diff
@@ -141,11 +145,11 @@ export class TypedMessageManager extends AbstractMessageManager<
    * @param req - The original request object possibly containing the origin.
    * @returns The id of the newly created TypedMessage.
    */
-  addUnapprovedMessage(
+  async addUnapprovedMessage(
```
Comment (Gudahtt) [inline packages/message-manager/src/TypedMessageManager.ts:148]: Looks like this is a breaking change (just noting this for the changelog)

**Pull request merged**